addons/web/static/src/webclient/company_service.js

```diff
@@ -84,43 +64,142 @@ export const companyService = {
+     * while logging always activates.
+     *
+     * @param {"toggle"|"loginto"} mode
+     * @param {string} companyId
+     */
+    setCompanies(mode, companyId) {
+        if (mode === "toggle") {
+            if (this.selectedCompaniesIds.includes(companyId)) {
+                this._deselectCompany(companyId);
+            } else {
+                this._selectCompany(companyId);
+            }
+        } else if (mode === "loginto") {
+            this.selectedCompaniesIds.splice(0, this.selectedCompaniesIds.length);
+            this._selectCompany(companyId);
+        }
+    }
+
+    /**
+     * Reload the browser to update the active companies to
+     * be equals to the selected companies.
+     */
+    applySelectedCompanies() {
+        const newCompanyIds = this.selectedCompaniesIds.length
+            ? this.selectedCompaniesIds
+            : [this.activeCompanyIds[0]];
+
+        const cidsHash = formatCompanyIds(newCompanyIds, CIDS_HASH_SEPARATOR);
+        this._router.pushState({ cids: cidsHash }, { lock: true });
+        cookie.set("cids", formatCompanyIds(newCompanyIds));
+        browser.setTimeout(() => browser.location.reload()); // history.pushState is a little async
+    }
+
+    /**
+     * Adds a company and each of its branches to
+     * the selected companies.
+     * @private
+     */
+    _selectCompany(companyId) {
+        if (!this.selectedCompaniesIds.includes(companyId)) {
+            this.selectedCompaniesIds.push(companyId);
+            this.getCompany(companyId).child_ids.forEach((companyId) =>
+                this._selectCompany(companyId)
+            );
+        }
+    }
+
+    /**
+     * Removes a company and each of its branches from
+     * the selected companies.
+     * @private
+     */
+    _deselectCompany(companyId) {
+        if (this.selectedCompaniesIds.includes(companyId)) {
+            this.selectedCompaniesIds.splice(this.selectedCompaniesIds.indexOf(companyId), 1);
+            this.getCompany(companyId).child_ids.forEach((companyId) =>
+                this._deselectCompany(companyId)
+            );
+        }
+    }
+
+    /**@private*/
+    _getCompanyIdsFromBrowser() {
+        let cids;
+        const hash = this._router.current.hash;
+        if ("cids" in hash) {
+            // backward compatibility s.t. old urls (still using "," as separator) keep working
+            // deprecated as of 17.0
+            let separator = CIDS_HASH_SEPARATOR;
+            if (typeof hash.cids === "string" && !hash.cids.includes(CIDS_HASH_SEPARATOR)) {
+                separator = ",";
+            }
+            cids = parseCompanyIds(hash.cids, separator);
+        } else if (cookie.get("cids")) {
+            cids = parseCompanyIds(cookie.get("cids"));
+        }
+        return cids || [];
+    }
+}
+
+export const companyService = {
+    dependencies: ["user", "router", "action"],
+    start(env, services) {
+        return new CompanyService(env, services);
```

Comment:
I'm not sure about the class, because now internals of the service are accessible in the service API (even tough they're private and should not be used). I think functions in the scope better separates the private concerns and the public API. Also, it makes the public API clearer when you debug as you only see what you can really access. That's a personal opinion, we can discuss about it if you disagree.

Reply:
Changed it to a function 👍